common/docker/client_image.go

```diff
@@ -0,0 +1,174 @@
+	"pkg.world.dev/world-cli/common/logger"
+)
+
+func (c *Client) pullImageIfNotExists(ctx context.Context, configuration service.Service) error {
+	fmt.Println(contextPrint("Pulling", "2", "image", "4", configuration.Image))
+
+	_, _, err := c.cli.ImageInspectWithRaw(ctx, configuration.Image)
+
+	// If image exists, return
+	if err == nil {
+		logger.Println("Image already exists", configuration.Image)
+		return nil
+	}
+
+	// If image does not exist, pull it
+	if client.IsErrNotFound(err) {
+		pullOptions := image.PullOptions{}
+
+		// Set platform if specified
+		if configuration.Platform.Architecture != "" && configuration.Platform.OS != "" {
+			pullOptions.Platform = fmt.Sprintf("%s/%s", configuration.Platform.OS, configuration.Platform.Architecture)
+		}
+
+		out, err := c.cli.ImagePull(ctx, configuration.Image, pullOptions)
+		if err != nil {
+			return err
+		}
+		defer out.Close()
+
+		return c.filterDockerPullOutput(out)
+	}
+
+	return err
+}
+
+func (c *Client) buildImage(ctx context.Context, dockerfile service.Dockerfile, imageName string) error {
+	fmt.Println(contextPrint("Building", "2", "image", "4", imageName))
+	buf := new(bytes.Buffer)
+	tw := tar.NewWriter(buf)
+	defer tw.Close()
+
+	// Add the Dockerfile to the tar archive
+	header := &tar.Header{
+		Name: "Dockerfile",
+		Size: int64(len(dockerfile.Script)),
+	}
+	if err := tw.WriteHeader(header); err != nil {
+		return err
+	}
+	if _, err := tw.Write([]byte(dockerfile.Script)); err != nil {
+		return err
+	}
+
+	// Add source code to the tar archive
+	if err := c.addFileToTarWriter(".", tw); err != nil {
+		return err
+	}
+
+	// Read the tar archive
+	tarReader := bytes.NewReader(buf.Bytes())
+
+	buildOptions := types.ImageBuildOptions{
+		Dockerfile: "Dockerfile",
+		Tags:       []string{imageName},
+		Target:     dockerfile.Target,
+	}
+
+	if c.cfg.BuildkitSupport {
+		buildOptions.Version = types.BuilderBuildKit
+	}
+
+	// Build the image
+	buildResponse, err := c.cli.ImageBuild(ctx, tarReader, buildOptions)
+	if err != nil {
+		return err
+	}
+	defer buildResponse.Body.Close()
+
+	// Print the build logs
+	err = c.filterDockerBuildOutput(buildResponse.Body)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) filterDockerBuildOutput(reader io.Reader) error {
+	decoder := json.NewDecoder(reader)
+	for {
+		var event map[string]interface{}
+		if err := decoder.Decode(&event); errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		// Filter the content
+		// Only print the stream and error fields
+		if val, ok := event["Stream"]; ok && val != "" {
+			fmt.Println(val)
+		} else if val, ok = event["error"]; ok && val != "" {
+			fmt.Println(contextPrint("Error:", "1", "", "1", val.(string)))
+		}
+		// else if val, ok = event["aux"]; ok {
+		// 	// Do nothing if build are using buildkit
+		// }
+	}
+	return nil
+}
+
+func (c *Client) filterDockerPullOutput(reader io.Reader) error { //nolint:gocognit
```

Comment:
similar to my comment above, i think logging the pull log after the fact (other than the error message, ofc) is not really useful; if any, we should be logging the progression of the image pull, otherwise user might think that the CLI is stuck.

Reply:
adjusted this on new func `pullImages` in `client_image.go`